### ISSUE TITLE
Nice work but some questions for this model

### ISSUE THREAD

**Issue by kelisiya** [zsyOAOA/InvSR#1]
When I run the demo to test whole_img , some blue pixels in the output image , can you solve this problem. And will support x2 sr ?

**Comment by zsyOAOA** [comment]
Yes, I know this issue. 

1. The images in the whole_img folder are borrowed from the face restoration project called CodeFormer. For face restoration, the related models are usually trained with a very heavy degradation pipeline, see CodeFormer.

2. Our InvSR model is trained with the degradation pipeline of RealEsrGAN, which is specifically designed for natural image super-resolution. Thus, our model is not robust enough to deal with these images for face restoration cases. I guess this issue can be solved by fine-tuning or retraining our model with the degradation pipeline for face restoration.    

3. I have tested our model on many low-quality images used in natural image super-resolution tasks and did not find a similar issue.

**Comment by zsyOAOA** [comment]
In addition, I do not train the x2 SR model.

**Comment by jacooooooooool** [comment]
hello, your project is very cool! , but "face reconstruction" is weak :( additionally the face reconstruction is unnatural which can be seen in magnification, especially eyes and mouth, denoising is good! , you explained why and I understand that it is not the fault of your project but the model used :) I would like to ask for a few tips:

1. can I use codeformer models? which are amazing and much faster when rebuilding faces?

2. is the TRAINING process of your project a process of "tuning" or retraining your model using the degradation pipeline for BETTER face restoration?

I would also like to add that the whole process works correctly on ubuntu 22 with nvidia GTX 1080Ti but it is about 10x slower than codeformer, while maintaining the same system and photo for processing

**Comment by zsyOAOA** [comment]
Thanks for your attention. I want to further clarify some tips on general image enhancement. @jacooooooooool 

1. A general image enhancement pipeline often involves two restoration models, one general enhancement model to handle the non-face areas and one specific face model for the face areas. The final results are blended according to the face landmarks after processing by these two models.  For example, to deal with general cases, the CodeFormer approach adopts RealESRGAN to enhance the non-face area and enhances the faces using the specifically trained model for faces (i.e., CodeFormer).

2. To deal with the general images in this project, we **do not** use another specific model to enhance the face areas. Thus, our method is weak in the general images with heavily degraded faces.

3.  As for the speed, on the one hand, our code will chop the images into small patches with a resolution of 128 x 128 and enhance each patch one by one. This chopping strategy aims to save GPU memory. On the other hand, our method employs a large-scale Stable Diffusion model than CodeFormer. Thus, it is slower than CodeFormer. 

4. To avoid  misunderstanding, I have removed the application of general image enhancement.

**Comment by kelisiya** [comment]
> Thanks for your attention. I want to further clarify some tips on general image enhancement. @jacooooooooool
> 
> 1. A general image enhancement pipeline often involves two restoration models, one general enhancement model to handle the non-face areas and one specific face model for the face areas. The final results are blended according to the face landmarks after processing by these two models.  For example, to deal with general cases, the CodeFormer approach adopts RealESRGAN to enhance the non-face area and enhances the faces using the specifically trained model for faces (i.e., CodeFormer).
> 2. To deal with the general images in this project, we **do not** use another specific model to enhance the face areas. Thus, our method is weak in the general images with heavily degraded faces.
> 3. As for the speed, on the one hand, our code will chop the images into small patches with a resolution of 128 x 128 and enhance each patch one by one. This chopping strategy aims to save GPU memory. On the other hand, our method employs a large-scale Stable Diffusion model than CodeFormer. Thus, it is slower than CodeFormer.
> 4. To avoid  misunderstanding, I have removed the application of general image enhancement.

Can you give me the Wechat ? I can send information to your designated email address . 
I am also a related researcher, this is my homepage kelisiya.github.io

**Comment by kelisiya** [comment]
> Thanks for your attention. I want to further clarify some tips on general image enhancement. @jacooooooooool
> 
> 1. A general image enhancement pipeline often involves two restoration models, one general enhancement model to handle the non-face areas and one specific face model for the face areas. The final results are blended according to the face landmarks after processing by these two models.  For example, to deal with general cases, the CodeFormer approach adopts RealESRGAN to enhance the non-face area and enhances the faces using the specifically trained model for faces (i.e., CodeFormer).
> 2. To deal with the general images in this project, we **do not** use another specific model to enhance the face areas. Thus, our method is weak in the general images with heavily degraded faces.
> 3. As for the speed, on the one hand, our code will chop the images into small patches with a resolution of 128 x 128 and enhance each patch one by one. This chopping strategy aims to save GPU memory. On the other hand, our method employs a large-scale Stable Diffusion model than CodeFormer. Thus, it is slower than CodeFormer.
> 4. To avoid  misunderstanding, I have removed the application of general image enhancement.

Will support bf16 or fp16 ?

**Comment by zsyOAOA** [comment]
To solve this issue, you should finetune our model using the degradation pipeline used in face restoration, e.g., https://github.com/zsyOAOA/DifFace/blob/master/configs/training/swinir_ffhq512.yaml#L25

**Comment by kelisiya** [comment]
> To solve this issue, you should finetune our model using the degradation pipeline used in face restoration, e.g., https://github.com/zsyOAOA/DifFace/blob/master/configs/training/swinir_ffhq512.yaml#L25

I will add relevant portrait data to fine-tune according to the current degradation, hoping to improve the noise estimation model regarding the problem of pixels predicted for some faces. If there is any relevant progress, I will further share it here

**Comment by iwanttostudyplease** [comment]
Hello landlord and author, have you encountered this problem?
![image](https://github.com/user-attachments/assets/6dbc0167-d866-4803-b077-adf93eeafcd4)
Can you help with that, thank you?

**Comment by kelisiya** [comment]
Regarding the further discussion on the appearance of blue pixel patch, in fact, by reasoning through the code, I have saved the noise, and it can be observed that the image before denoising already has noticeable blue blocks, making it challenging to completely remove them even with sd-turbo(0.25 noise). The problem may not solely arise from the model's lack of fitting. If you have the time, could you please revisit your code to see if this problem can be resolved or provide a way to detect the blue blocks? I am also working on optimizing this problem, be it through training or coding .
![088abc17-954e-41ad-b071-55e57277eba8](https://github.com/user-attachments/assets/967b1a5f-b9da-4dcf-836a-d41571e23f94)

**Comment by zsyOAOA** [comment]
@kelisiya We have uploaded a new [model](https://huggingface.co/OAOA/InvSR/blob/main/noise_predictor_sd_turbo_v5_diftune.pth) in huggingface. Please download this model and try again via: 

python inference_invsr.py -i [image folder/image path] -o [result folder] --num_steps 1 --timesteps 200 --started_ckpt_path download_path --chopping_size 256

**Comment by kelisiya** [comment]
> @kelisiya We have uploaded a new [model](https://huggingface.co/OAOA/InvSR/blob/main/noise_predictor_sd_turbo_v5_diftune.pth) in huggingface. Please download this model and try again via:
> 
> python inference_invsr.py -i [image folder/image path] -o [result folder] --num_steps 1 --timesteps 200 --started_ckpt_path download_path --chopping_size 256

Can you share the experiment about how to finetune the model ?

**Comment by zsyOAOA** [comment]
We first train a model using only the diffusion loss (L2), and then initialize from this model by adding GAN and LPIPS loss. 

Equivalently, one can directly adjust the config by increasing the hyper-parameters: train.dis_init_iterations.

**Comment by kelisiya** [comment]
> We first train a model using only the diffusion loss (L2), and then initialize from this model by adding GAN and LPIPS loss.
> 
> Equivalently, one can directly adjust the config by increasing the hyper-parameters: train.dis_init_iterations.

In this new model , use more dataset or only finetune ?

**Comment by zsyOAOA** [comment]
> > We first train a model using only the diffusion loss (L2), and then initialize from this model by adding GAN and LPIPS loss.
> > Equivalently, one can directly adjust the config by increasing the hyper-parameters: train.dis_init_iterations.
> 
> In this new model , use more dataset or only finetune ?

No, we do not use more datasets, just only with more training time. @kelisiya

**Comment by kelisiya** [comment]
> > > We first train a model using only the diffusion loss (L2), and then initialize from this model by adding GAN and LPIPS loss.
> > > Equivalently, one can directly adjust the config by increasing the hyper-parameters: train.dis_init_iterations.
> > 
> > 
> > In this new model , use more dataset or only finetune ?
> 
> No, we do not use more datasets, just only with more training time. @kelisiya

Thank! Orz,And if I want to fine-tune this model, do I need to further adjust settings such as loss or learning rate? How can I keep it consistent with training your new model? Thanks again for your prompt responses be lower and any

**Comment by kelisiya** [comment]
> > > We first train a model using only the diffusion loss (L2), and then initialize from this model by adding GAN and LPIPS loss.
> > > Equivalently, one can directly adjust the config by increasing the hyper-parameters: train.dis_init_iterations.
> > 
> > 
> > In this new model , use more dataset or only finetune ?
> 
> No, we do not use more datasets, just only with more training time. @kelisiya

Orz thanks ! If I want to fine-tune this model, do I need to further adjust settings such as loss or learning rate? How can I keep it consistent with training your new model? Thanks again for your responses ~

**Comment by zsyOAOA** [comment]
You have to keep the config for model.params, which controls the architecture of the noise predictor, and other configs could be freely adjusted. @kelisiya

**Comment by kelisiya** [comment]
> You have to keep the config for model.params, which controls the architecture of the noise predictor, and other configs could be freely adjusted. @kelisiya

How can I use chopping size 256 with less GPU memory likes 24G RAM ? Do I need to change the slicing algorithm?

**Comment by kelisiya** [comment]
> You have to keep the config for model.params, which controls the architecture of the noise predictor, and other configs could be freely adjusted. @kelisiya

Can you share the weights of a discriminator with me? I found that using pretraining to fine-tuning ,but the test image sharpen when test the model during training. I would like to further verify if this is caused by retraining the discriminator. Thank you.
![output](https://github.com/user-attachments/assets/dffa5f81-4bd2-47f2-8abc-86575b23f001)

**Comment by zsyOAOA** [comment]
> > You have to keep the config for model.params, which controls the architecture of the noise predictor, and other configs could be freely adjusted. @kelisiya
> 
> Can you share the weights of a discriminator with me? I found that using pretraining to fine-tuning ,but the test image sharpen when test the model during training. I would like to further verify if this is caused by retraining the discriminator. Thank you. ![output](https://private-user-images.githubusercontent.com/23169766/400674618-dffa5f81-4bd2-47f2-8abc-86575b23f001.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYyNDAyMzMsIm5iZiI6MTczNjIzOTkzMywicGF0aCI6Ii8yMzE2OTc2Ni80MDA2NzQ2MTgtZGZmYTVmODEtNGJkMi00N2YyLThhYmMtODY1NzViMjNmMDAxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA3VDA4NTIxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjMzFmNjExYzc1ODVkZTVlMzg3MTAwMDhkZWQzZmJlYmI4NjIzOTVjZjJlMDNjMDkwMDU3ZmExMTUwYjM4NzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.agXieNcS5CcxMZCw-Fw_lgT-cIOUyCtJW9vuGjqeTpc)

I have uploaded the trained [discriminator](https://huggingface.co/OAOA/InvSR/blob/main/discriminator_sd_turbo_v5.pth) for your information.

**Comment by laihongphuc** [comment]
> Regarding the further discussion on the appearance of blue pixel patch, in fact, by reasoning through the code, I have saved the noise, and it can be observed that the image before denoising already has noticeable blue blocks, making it challenging to completely remove them even with sd-turbo(0.25 noise). The problem may not solely arise from the model's lack of fitting. If you have the time, could you please revisit your code to see if this problem can be resolved or provide a way to detect the blue blocks? I am also working on optimizing this problem, be it through training or coding . ![088abc17-954e-41ad-b071-55e57277eba8](https://github.com/user-attachments/assets/967b1a5f-b9da-4dcf-836a-d41571e23f94) patch inference update ![image](https://github.com/user-attachments/assets/b10f4392-e4bb-427b-a2e3-5b4fb4df060b)

Hi @kelisiya , I’m curious about your solution to the “blue blocks” artifact issue. Were you able to completely eliminate it? I’ve only encountered this problem with the SD-Turbo model.
